File: 115 Distinct Subsequences.c
/*
115. Distinct Subsequences
https://leetcode.com/problems/distinct-subsequences/

Given a string S and a string T, count the number of distinct subsequences of S which equals T.

A subsequence of a string is a new string which is formed from the original string by deleting some (can be none)
of the characters without disturbing the relative positions of the remaining characters.
(ie, "ACE" is a subsequence of "ABCDE" while "AEC" is not).

Example 1:

Input: S = "rabbbit", T = "rabbit"
Output: 3
Explanation:

As shown below, there are 3 ways you can generate "rabbit" from S.
(The caret symbol ^ means the chosen letters)

rabbbit
^^^^ ^^
rabbbit
^^ ^^^^
rabbbit
^^^ ^^^
Example 2:

Input: S = "babgbag", T = "bag"
Output: 5
Explanation:

As shown below, there are 5 ways you can generate "bag" from S.
(The caret symbol ^ means the chosen letters)

babgbag
^^ ^
babgbag
^^    ^
babgbag
^    ^^
babgbag
  ^  ^^
babgbag
    ^^^
*/
/*
Let's say i is iterator of s, j is iterator of t,
then dp[i][j] = dp[i-1][j] + dp[i-1][j-1] if dp[i] == dp[j] else dp[i-1][j]
*/
int numDistinct(char * s, char * t)
{
    uint64_t i,j, upper, upperLeft;
    uint64_t* row;
    int ls = (int)strlen(s);
    int lt = (int)strlen(t);
    
    if (ls == 0 || lt > ls) return 0;
	if (lt == 0) return 1;
    
    row = (uint64_t*)calloc(lt, sizeof(uint64_t));
    
    for (i = 0; i < ls; i++)
    {
        // upperLeft is at least 1, which means if t is empty, therer is 1 distinct subsequence
        upperLeft = 1u;
        for (j = 0; j < lt; j++)
        {
            upper = row[j];
            row[j] = (s[i] == t[j] ? upperLeft + upper : upper);
            upperLeft = upper;
        }
    }
    i = row[lt-1];
    free(row);
    return (int)i;
}

File: 115 Distinct Subsequences.py
'''
115. Distinct Subsequences
https://leetcode.com/problems/distinct-subsequences/

Given two strings s and t, return the number of distinct subsequences of s which equals t.
The test cases are generated so that the answer fits on a 32-bit signed integer.

Example 1:

Input: s = "rabbbit", t = "rabbit"
Output: 3
Explanation:
As shown below, there are 3 ways you can generate "rabbit" from s.
rabbbit
rabbbit
rabbbit

Example 2:

Input: s = "babgbag", t = "bag"
Output: 5
Explanation:
As shown below, there are 5 ways you can generate "bag" from s.
babgbag
babgbag
babgbag
babgbag
babgbag

Constraints:
    1 <= s.length, t.length <= 1000
    s and t consist of English letters.
'''
class Solution:
    def numDistinct(self, source: str, target: str) -> int:
        # dp[s][t] is the distinct count of src[:s] to tag[:t]
        dp = [[0] * (len(target)+1) for _ in range(len(source)+1)]
 
        # If target is empty, there is always 1 distinct subsequence.
        # This includes any source[:] and empty target case.
        for s in range(len(source)+1):
            dp[s][0] = 1

        for s in range(len(source)):
            for t in range(len(target)):
                if source[s] == target[t]:
                    dp[s+1][t+1] = dp[s][t+1] + dp[s][t]
                else:
                    dp[s+1][t+1] = dp[s][t+1]
                    
        return dp[-1][-1]
